scene/animation/animation_tree.cpp

```diff
@@ -2203,6 +2208,80 @@ real_t AnimationTree::get_connection_activity(const StringName &p_path, int p_co
 	return (*activity)[p_connection].activity;
 }
 
+void AnimationTree::set_record_buffer_max_size(uint32_t p_max_size) {
+	record_buffer_max_size = p_max_size;
+
+	record_buffer.adjust_capacity(record_buffer_max_size);
+}
+
+void AnimationTree::resize_record_buffer(int p_size) {
+	record_buffer.resize(p_size);
+}
+
+void AnimationTree::discard_latest_records(uint32_t p_count) {
+	ERR_FAIL_COND(p_count > static_cast<uint32_t>(record_buffer.size()));
+
+	for (uint32_t i = 0; i < p_count; ++i) {
+		record_buffer.pop_front();
+	}
+}
+
+void AnimationTree::set_recording_rate(real_t p_rate) {
+	recording_rate = p_rate;
+	real_t min_rate = 1.0 / Engine::get_singleton()->get_frames_per_second();
+	if (recording_rate > min_rate) {
+		recording_rate = min_rate;
+	}
+}
+
+void AnimationTree::save_record(bool p_save_invalid_reasons) {
+	Record new_record;
+	new_record.state_animation_states.reserve(state.animation_states.size());
+	for (const AnimationNode::AnimationState &E : state.animation_states) {
+		new_record.state_animation_states.push_back(E);
+	}
+	new_record.state_last_pass = state.last_pass;
+	new_record.state_valid = state.valid;
+	new_record.process_pass = process_pass;
```

Comment:
I think these three variables are not necessary. Instead, I assume you need some way to avoid doing double processing after the call to `load_record()`.

Reply:
`state.last_pass` will be overrided in `_process_graph`, it can't be removed.

It seem `state.last_pass` and `process_pass` only be used to get activity in editor plugin,
`state.valid` will be overrided in `_process_graph`, too, and it is not be expoed to GDScript, It seem only work with `state.state_invalid_reasons` in editor.

In conclusion, I remove `state.last_pass`, and handel `process_pass`, `state.valid` and `state.state_invalid_reasons` only in tool build.

Reply:
I wanted to say that instead of recording this, it is necessary to set it up again after retrieving the parameter list with `load_record()`.

Reply:
Sorry, I still not clearfy what do you means.
I have changed the retrieving order of `Record`'s members in `load_record()`. Please review it.